### DISCUSSION TITLE
Switching to microk8s - 6 Months Later | benbrougher.tech

### DISCUSSION THREAD

**Discussion by giscus** [moutansos/benbrougher-tech#55] [Comments]
# Switching to microk8s - 6 Months Later | benbrougher.tech

How the move from kubeadm to microk8s has worked out over the last 6 months

https://benbrougher.tech/posts/microk8s-6-months-later/

**Comment by cppservergit** [comment]
Nice article, thank you for sharing your experience, I use MicroK8s over LXD, this way I can test HA cluster on a single machine (baremetal or VM), in order to solve the DNS problem in LXD I disabled IPv6:

lxc network set lxdbr0 ipv6.address=none

This way my k8s nodes running inside LXC containers can see all that is visible to my host.

LXD with MicroK8s and a native load balancer on the host machine, like HAProxy, proved to be a poweful combination to mount a cloud-like lab, it is fast to add nodes to LXD and then to the MicroK8s cluster, as well as updating HAProxy and reloading on-the-fly its configuration.

**Comment by cppservergit** [comment]
Another relevant point for the "Cons" would be the high CPU consumption by the dqlite process, even when there is no load on the Pods, it is certainly a known issue, and Canonical made some fixes for 1.27 (released a few days ago) but I still see the same high cpu mark for dqlite

**Comment by moutansos** [comment]
I'll have to give LXD/LXC a look. I have heard of them before but haven't taken a look to see really what they are and how they work, but that seems like an awesome way to get to play with HA on a single machine. I've seen HA/non-HA clusters behave very differently in the wild so testing/learning in HA always seems to be a better experience to me. And yeah I definitely see that dqlite process bubble up quite often. With the machines that I tend to use it's not much of an issue, especially considering our workloads aren't super high utilization. The company I work for owns their infrastructure so we tend to over-spec our VMs since we've got the space and we've already paid for it anyways.

**Reply by cppservergit** [reply]
When MicroK8s is in the non-HA mode it won't use Calico as the network manager, but Flannel, won't support NetworkPolicy for instance in non-ha mode, but it will support it in HA mode with Calico. I have observed that CPU consumption patterns vary a lot from VMs provisioned on desktop computers to VMs provisioned in server-class machines, like those I use on Azure. My current lab is Linux baremetal on a mini-pc (fits on your hand palm) with PCIe 4th gen, 32GB RAM and Ryzen 7, I partition this mini-pc into several light Linux native containers using LXD, this way I can play HA like in the cloud, and then run stress test from my PC with 10.000 concurrent users, I use HAProxy in front of my MicroK8s cluster to distribute the load to the nginx Ingresses

[In case you are curious about this setup](https://github.com/cppservergit/cppserver-docs/blob/main/networkpolicy.md)

Are you gonna upgrade the cluster to 1.27? They said they fixed several problems in dqlite, I still have to run the bombardment on my cluster to see how it does behave.

On baremetal Linux you can also use multipass to provision VMs, which us very nice and cloud-like experience, but then qemu-system-x86 runs higher than 100%, they haven't fixed that yet.

Regards,
Martin

**Reply by moutansos** [reply]
The mini-pc route is definitely a lot more space/cost/energy efficient than my current home lab solution. I've got VMs running on an old PowerEdge server and then some on my TrueNAS box, so I can take one of them down without taking the whole cluster down (it's not really HA when you have power outages on a regular basis 😄). 

I am definitely planning on upgrading both the work an home clusters to 1.27. I try to keep at least within a month of a Kubernetes release, just to keep things patched and up to date. I'll probably do my home lab in the next week and see how that goes. I definitely wouldn't complain about performance improvements especially on my lower spec'd VMs. I still have to go through the changelog and make sure I won't be breaking anything in the upgrade though.
